### ISSUE TITLE
Generated clientset can't watch CustomResources

### ISSUE THREAD

**Issue by jsafrane** [kubernetes/kubernetes#67602]
I am experimenting with CRD is `k8s.io/api`. I put my `types.go` to `staging/src/k8s.io/api/csi` and run `hack/update-codegen.sh`*. Such generated client(s) are part of the usual `clientset.Interface` with core Kubernetes APIs.

In kubelet and controller-manager, I can use the clientset to list my CRs  without any issues:

```go
client.Csi().CSIDrivers().List(meta.ListOptions{})
```

In kubelet and controller-manager, I can **not** use the clientset to **watch** CRs:
```go
w, err = client.Csi().CSIDrivers().Watch(meta.ListOptions{ResourceVersion: "0"})
if err != nil {
	glog.Infof("Watch error: %s", err)
	return
}
ch := w.ResultChan()
for {
	obj, ok := <-ch
	glog.Infof("watch got ok: %t, obj: %+v", ok, obj)
	if !ok {
		glog.Infof("watch closed")
		break
	}
}
```

`Watch()` returns just fine. But the returned channel `ch` closes immediately with this message in log:
```
E0820 14:55:10.279927   13837 streamwatcher.go:109] Unable to decode an event from the watch stream: got short buffer with n=0, base=851, cap=2688
```

It seems to be related to timing, sometimes the channel closes just like above, sometimes it's open, but never delivers any event, even if I create/delete CRs.

Everything gets working if I run controller-manager with `--kube-api-content-type="application/json"`, the channel reports all events as expected. It must be something with protobuf used for most of the API and json used for CRDs.

*) I had to hack `hack/update-codegen.sh` to work for API in `staging/src/k8s.io/api` without its internal counterpart in `pkg/apis`. I'll fill a separate issue/PR for it.

**Environment**:
- Kubernetes version (use `kubectl version`): today's master,
```
Client Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.0-alpha.0.301+ae5070d80ad248-dirty", GitCommit:"ae5070d80ad24841a876b964faed2c370dc91e04", GitTreeState:"dirty", BuildDate:"2018-08-20T14:55:19Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.0-alpha.0.301+0ff6f78edd315a-dirty", GitCommit:"0ff6f78edd315a6bb5a02f66c6ca997f77d51693", GitTreeState:"dirty", BuildDate:"2018-08-20T11:56:04Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: none, hack/local-up-cluster.sh
- OS (e.g. from /etc/os-release): centos 7

**Comment by jsafrane** [comment]
/sig api-machinery
cc @deads2k, @smarterclayton, @nikhita

**Comment by saad-ali** [comment]
CC @caesarxuchao

**Comment by nikhita** [comment]
/kind bug
/area custom-resources

@jsafrane by any chance, do you have these changes pushed? easier to debug/fix that way.

**Comment by jsafrane** [comment]
@nikhita, I tried to prepare an integration test, in https://github.com/kubernetes/kubernetes/pull/67657

I noticed that even `kubeclient.CsiV1alpha1().CSIDrivers().Create(driver)` failed with "object *v1alpha1.CSIDriver does not implement the protobuf marshalling interface and cannot be encoded to a protobuf message", I did not get to `Watch(()` with protobuf.

**Comment by josiahbjorgaard** [comment]
Bug triage for 1.17 here. This issue has been open for a significant amount of time and since it is tagged for milestone 1.17, we want to let you know that the 1.17 code freeze is coming in less than one month on Nov. 18th. Will this issue be resolved before then?

**Comment by josiahbjorgaard** [comment]
Hi, checking in to see if this issue intended to be resolved before code freeze on Nov. 14?

**Comment by liggitt** [comment]
This was due to the kube-controller-manager specifically pinning its supported content types to protobuf. It needed to indicate it accepted proto or json. Fixed by #77817, #84692, #85076,
